notes/2018-01-22-packaging.md

```diff
@@ -0,0 +1,104 @@
+* If Radicle app frontends that are written in radicle the interpreter also needs
+  to locate Radicle source files to be able to import them.
+
+In addition to installed packages the location mechanism needs to work for
+development builds (`stack build`), local tests, and CI tests.
+
+To implement file location of subcommands we are going to use Cabal’s
+[`Paths_radicle`][cabal-paths] module. `getBinDir` will return the folder
+containing all subcommands.
+
+File location for Radicle source files will be implemented by reading the
+`RADPATH` environment variable
+
+For local development and testing this mechanisms work out of the box. For
+packages the `rad` command will be a script with the following content.
+
+```
+#!/bin/sh
+radicle_bin=/usr/lib/radicle/bin
+RADPATH=/usr/lib/radicle/modules
+exec $radicle_bin/rad "@$"
+```
+
+Setting `radicle_bin` will determine the return value of `getBinDir`. Setting
+`RADPATH` will enable other commands to pick up Radicle source files.
+
+[cabal-paths]: https://cabal.readthedocs.io/en/latest/developing-packages.html#accessing-data-files-from-package-code
+
+Dependencies
+------------
+
+Radicle depends on the IPFS daemon. Packages must either register the IPFS
+daemon as a dependency or distribute the `ipfs` binary with them.
+
+Services
+--------
+
+For Radicle to work two services need to be running: the Radicle daemon and the
+IPFS daemon. For Debian in Pacman we will provide a Systemd Unit file for the
+Radicle daemon as well as installation instructions that include extra steps to
+enable the Radicle daemon service.
+
+The Radicle daemon Systemd Unit has a dependency on the IPFS Systemd Unit. If we
+distribute IPFS with the package we also need to add a Systemd Unit file for
+IPFS.
+
+We will investigate whether Brew provides a similar service management system
+
+Continuous Delivery
+-------------------
+
+Packages will be released using a release pipeline in CI. The release pipeline
+will build the packages, create a Github release with the appropriate name and
+description, and attach the packages to the Github release. The release pipeline
+will be triggered by Git tags.
+
+To build Brew packages we will need to run the build pipeline on MacOS. Google
+Cloud Build does not support MacOS but Travis and Circle do.
+
+Release pipeline
+----------------
+
+This is how the release pipeline would look for a Debian package. For a Pacman
+package only minor modifications would need to be made.
+
+1. `stack build -- --flag radicle:release`. The flag disables some targets,
+   enforces static linking and `-O3`.
+2. Move all binaries built with stack to `$DEST/usr/lib/radicle/bin`
+3. Move all Radicle sources files to `$DEST/usr/lib/radicle/modules`
+4. Create `$DEST/usr/bin/rad` wrapper script as mentioned above.
+5. Create a `deb` package using [`fpm`][fpm]. `fpm` takes the Systemd file as a CLI option.
+6. Upload the `deb` file to Github releases
+
```

Comment:
GPG Signatures?

Reply:
Any suggestions how to sign the release artifacts?

Reply:
If we make releases from a CD pipeline, it boils down to trusting the build machine. Perhaps releases shouldn't be cut from a SaaS provider?

Reply:
I’ll leave this as an open question for now. I added notes in a new section.